platform/core/src/core.rs

```diff
@@ -22,44 +31,60 @@ pub enum PlatformResponse {
 }
 
 pub trait PlatformHookRequest: Sync + Send {
-    fn perform_request(&self, request: PlatformRequest) -> Result<PlatformResponse, String>;
+    fn perform_request(&self, request: PlatformRequest) -> Result<PlatformResponse, CoreFailure>;
 }
 
+const PREFERENCES_DB: &str = "preferences_db";
+
 // The Core would act as a composition root, containing any global configuration
 // to be shared between components, as well managing components lifetime.
-struct Core {
+struct Core<'a> {
     language_code: String,
     #[allow(dead_code)]
+    preferences: Arc<Preferences>,
+    user_session: Mutex<UserSession<'a>>,
```

Comment:
Why Mutex and not Arc?

Reply:
`Mutex` and `Arc` are different things. `Arc` is used for reference counting, `Mutex` to allow internal mutability(in this case).

In this case `Mutex` is used since UserSession needs to be mutated, and it should happen in a thread safe manner.

Reply:
Is the reference to UserSession being mutated? Or the object itself being mutated?

Reply:
The object itself is being mutated.